### ISSUE TITLE
Reduce necessary DLLs footprint

### ISSUE THREAD

**Issue by forestbat** [dotnet/coreclr#25153]
![图片](https://user-images.githubusercontent.com/23413915/59445826-e1d74000-8e32-11e9-97f0-3da5f6fe1731.png)
Let's see this picture which is a hello world app trimmed by IL linker。
We can see System.Private.CoreLib is the biggest，about 10MB，and some necessary dlls such as coreclr，mscordbi are also a bit big，so I think it's important to reduce their size or split them to smaller pieces.

**Comment by AaronRobinsonMSFT** [comment]
cc @swaroop-sridhar @jeffschwMSFT @vitek-karas

**Comment by swaroop-sridhar** [comment]
CC: @sbomer

**Comment by swaroop-sridhar** [comment]
Related issue: #24397

**Comment by vitek-karas** [comment]
I just want to point out that this automatically assumes that self-contained app is necessary. It's very possible that in your scenario that is the case, but maybe not. Note that framework dependent apps are "MUCH" smaller - no amount of ILLinker will be able to get the app to the size of framework dependent app.

**Comment by cup** [comment]
@vitek-karas you posted the same type of comment in my thread:

https://github.com/dotnet/coreclr/issues/24397#issuecomment-489396948

I didn’t appreciate it then, and I dont appreciate it now. I dont think that
type of "political" comment is appropriate. You may find this build style
distasteful, but it has legitimate and real world uses and situations that would
benefit from it.

Could you please refrain from these off topic comments?

**Comment by forestbat** [comment]
@vitek-karas Framework dependent apps still need cost at least 70MB space on disk，doesn't it？
@AaronRobinsonMSFT @swaroop-sridhar @cup @sbomer @jeffschwMSFT 
IL linker works like -jlint in Java9，however the largest piece of a trimmed JRE is jbase——it's 17MB！With CoreCLR's changing，Corelib may be larger and larger until it become a new “jbase”，it's not a good future。
And I'm a player of Minecraft Java Edition，I‘m bored of telling newcomer of the game how to download JRE or JDK from Oracle.com/java.com，especially today we must register oracle account or search openjdk——in many scenes consumers are foolish，they don't want to care how to download a big dependence，what they want is a simple single file。
Although this is a big project，however it has benefit for future。Please think it over。

**Comment by AaronRobinsonMSFT** [comment]
@forestbat Yes the disk cost will still be 70 MB. @vitek-karas's point is that it can be a one time cost shared across all potential applications. If the machine will have multiple .NET applications using the same framework then the cost can be shared across all of them since each one will use the same CoreCLR. If however there is only a single program using CoreCLR then the cost isn't shared and much higher ratio per application.

In the scenario mentioned, a complete standalone to reduce users from having to download prerequisites the shared framework isn't as obvious. Thank you for bringing up this example it is helpful to have concrete scenarios.9

**Comment by forestbat** [comment]
![图片](https://user-images.githubusercontent.com/23413915/59547712-d50f3500-8f75-11e9-954c-3f4849d739ca.png)
And this is WPF app which has been trimmed。
![图片](https://user-images.githubusercontent.com/23413915/59547761-9ded5380-8f76-11e9-812b-6c6f7ad724c6.png)
This is ASP.NET Core app which has been trimmed。

**Comment by onovotny** [comment]
I would love to see a significant reduction in the final trimmed size as well. Not sure what the right mechanism is yet (more aggressive trimming of the private corlib, splitting into different modules, something else?). This also plays into the single-exe AOT story.

**Comment by forestbat** [comment]
@onovotny I think splitting into smaller modules and reduce footprint of every dll are better than trimming dlls more aggressively，because trimming a full assembly will probably generate no-standard dlls and bring some trouble。
AOT such as Mono AOT or CoreRT can largely reduce the assembly，however it will lose syntax such as generic and we probably wait for compile like C++。

**Comment by tomrus88** [comment]
I'm still not sure why they have 2 copies of mscordaccore.dll in output (mscordaccore.dll and mscordaccore_amd64_amd64_4.700.19.31301.dll). Both dll's are identical content wise, just different name...
Easy size reduction by 1.2MB by removing one of them...

**Comment by 0xd4d** [comment]
@tomrus88 most people who trim their apps probably don't want any of the debugger related files at all so they shouldn't even be there. DiaSymreader shouldn't be needed either since anyone who trims their app surely won't include the PDB files, and even if they did, they'd be portable PDBs not Windows PDBs.

**Comment by cup** [comment]
https://github.com/mono/linker/issues/607

**Comment by forestbat** [comment]
@AaronRobinsonMSFT Will this issue be added in 3.0 next milestone？

**Comment by AaronRobinsonMSFT** [comment]
@forestbat This issue will continue to get focus in future milestones. Both @sbomer and @swaroop-sridhar are actively working in this area to improve ILLinker scenarios to reduce managed binaries size in self-contained applications. One of them would have to comment on approaches being taken to reduce the native binary size. The question of whether we ever satisfy everyone's desires on size is another matter, but please continue to file specific issues and they will get prioritized against other issues.

@tomrus88 and @0xd4d both mentioned binaries that could be removed. @jeffschwMSFT, @sbomer, or @swaroop-sridhar do any of you know who from the diagnostic side of things is working on removing DAC and symbol reading binaries from the self contained scenario?

**Comment by sbomer** [comment]
@mikem8361 would you mind commenting on the question about removing diagnostics binaries?

**Comment by mikem8361** [comment]
sos.dll and sos_*.dll can be removed. SOS is now shipped/installed out of band.

The DAC (mscoredaccore.dll) and DBI (mscordbi.dll) DO need to be included to support managed debugging.

**Comment by sbomer** [comment]
Thanks! Is someone on the diagnostics team working on removing the unnecessary dlls from the runtime packages?

**Comment by cup** [comment]
I have a question - sorry if it is naive - but look at this:

https://github.com/activescott/lessmsi/releases

extracted the total size is 1.5 MB, and it is C#. how come the smallest I can get is 29 MB:

https://github.com/dotnet/coreclr/issues/24397#issuecomment-501519761

**Comment by mikem8361** [comment]
I don’t think we were aware that the diagnostics team was supposed to do anything.

@tommcdon

**Comment by forestbat** [comment]
@cup Of course it's native.
https://github.com/dotnet/coreclr/issues/24397#issuecomment-489391079
And 1.5M is the limit of a C# app，in this size coreclr won't exist and C# will lose almost all attributes of itself,I think MS want a small but contains many "normal" function of C# and CLR.
However I think a hello world app should be at most 10MB.

**Comment by tomrus88** [comment]
@cup It's C# on .NET Framework. You can do the same on .NET Core but installing .NET Core runtime on target system manually and publishing your app as framework dependent... It will have size similar to .NET Framework app size, very small.

**Comment by cup** [comment]
@tomrus88 where is the `dotnet.exe` for .NET Framework? where is that repo?

**Comment by tomrus88** [comment]
@cuo .NET Framework doesn't have or need dotnet.exe.

**Comment by cup** [comment]
@tomrus88 sorry I mean whats the equivalent

**Comment by vitek-karas** [comment]
In .NET Framework one of the jobs of the `dotnet.exe` is handled by the OS. Windows "knows" about .NET applications and will ask the installed .NET Framework to run the app if a managed executable is started as a process. This means that there was a small component which was part of the OS. We didn't want to do that in .NET Core as it's hard to service such component (and there's really no counterpart on Linux/Mac anyway).

The other things `dotnet.exe` can do, like `dotnet build` and similar SDK operations didn't have a counterpart on .NET Framework (I guess msbuild, but it's not the same). Note that you can use `dotnet build` and other SDK commands to work with .NET Framework projects now though.

**Comment by forestbat** [comment]
About 1 month passed，will it has some change in .NET 5？

**Comment by vitek-karas** [comment]
We've been focusing on finishing 3.0. There are no firm plans for 5 yet, but I think making single-file smaller will be part of it.

**Comment by forestbat** [comment]
#26175 I saw this issue，will structure of dotnet's packages change？

**Comment by jeffschwMSFT** [comment]
@forestbat feel free to comment on the main thread, there are no anticipated changes to the packages due to the repo consolidation.

**Comment by yyjdelete** [comment]
https://github.com/dotnet/coreclr/issues/25153#issuecomment-502396024
Is anyone still active on this and know why something like `mscordaccore_amd64_amd64_4.700.19.45506.dll` is still needed?
It's the same as `mscordaccore.dll`, and seems not be referenced by any files in the publish folder.

**Comment by jkotas** [comment]
> mscordaccore_amd64_amd64_4.700.19.45506.dll is still needed?

It is needed for dump debugging. It should not be needed in published self-contained apps. cc @dotnet/dotnet-diag 

Note that there are number of other files like this. For example, none of the `*.a` files in the self-contained apps published for Linux are needed.

@sbomer Do you have thoughts on what would be the best strategy to get these one-off unmanaged files trimmed from the self-contained apps?

**Comment by 0xd4d** [comment]
All debugger related files don't need to be included. They just add space and almost no-one needs them. Those that need them can easily find those files.

**Comment by jeffschwMSFT** [comment]
> @sbomer Do you have thoughts on what would be the best strategy to get these one-off unmanaged files trimmed from the self-contained apps?

Trimming native components did not make .NET Core 3.  But it is part of our plans for an upcoming release.

**Comment by cup** [comment]
Another option is Mono:

https://mono-project.com

Example:

~~~
C:\mono-6.4.0.198-x64-0\SourceDir\Mono> type sun-mon.cs
using System;
class Program {
   static void Main() {
      Console.WriteLine("sunday monday");
   }
}

C:\mono-6.4.0.198-x64-0\SourceDir\Mono> bin\mcs.bat sun-mon.cs
C:\mono-6.4.0.198-x64-0\SourceDir\Mono> sun-mon.exe
sunday monday

C:\mono-6.4.0.198-x64-0\SourceDir\Mono> dir sun-mon.exe
2019-11-10  09:28 AM             3,072 sun-mon.exe
~~~
